Feature/contracts transpiler

ylv-io opened the pull request:
The goal of this PR is to create a library of capable to transpile Solidity smart contracts to their upgradable version compatible with [OpenZeppelin SDK](https://openzeppelin.com/sdk/).

It is archived by transpiling Solidity source code. For example a file `Simple.sol` is transpiled to a file `SimpleUpgradable.sol` which is exactly the same except the following differences:
* Constructor is removed and replaced with `initialize` method with the same parameters.
* Contract is inherited from `[Initializable](https://github.com/OpenZeppelin/openzeppelin-sdk/blob/master/packages/lib/contracts/Initializable.sol)` in order to prevent initialization more than once.
* Extra storage gap is added at the end of the contract to enable future upgrades

For example:
```Solidity
contract Simple {
    uint256 public count;
    uint256 private local = 564;
    string private hello = "hello";
    bool public test = true;
    uint256 constant const = 32**22 + 8;

    constructor(uint256 num) public {
        count = num;
    }

    function getHello() public view returns (string memory) {
        return hello;
    }
}
```
would be turned into:
```Solidity
contract SimpleUpgradable is Initializable {
    uint256 public count;
    uint256 private local ;
    string private hello ;
    bool public test ;
    uint256 constant const = 32**22 + 8;

    function initialize(uint256 num) external initializer {
        __init(true, num);
    }

    function __init(bool callChain, uint256 num) internal {
        local = 564;
        hello = "hello";
        test = true;
        count = num;
    }

    function getHello() public view returns (string memory) {
        return hello;
    }
}
```

ylv-io commented:
**Public API for Transpiler**
Transpiler exports only one public API method `transpileContracts`.

```Typescript
export function transpileContracts(
  contracts: string[],  // array of contract names to transpile
  artifacts: Artifact[] // all required artifacts to do the job as json objects
):OutputFile[] // will return an array of transpilied files
```

```Typescript
export interface OutputFile {
  fileName: string; // an original file, for example GLDToken.sol
  source: string;  // a resulted source code of a file as a string
  path: string; // a relative path of a file at file system, for example ./contracts/__upgradable__/GLDTokenUpgradable.sol
  contracts: string[]; // an array of contracts names as strings contained by a file for example ["GLDToken"] 
}
```

To clarify on the path field. It will contain a relative path to a file staring with `./contracts/__upgradable__`.  `__upgradable__` is a special directory inside `contracts` directory containing all the transpilied contracts. All the required import directives will be patched and expect corresponding files to be located in the relative locations inside contracts folder for example:

![image](https://user-images.githubusercontent.com/3285000/71979145-f1e3c080-3257-11ea-97b1-d76ed3dc43e2.png)


```Solidity
import "./@openzeppelin/contracts/token/ERC20/ERC20Upgradable.sol";
import "./@openzeppelin/contracts/token/ERC20/ERC20DetailedUpgradable.sol";
import "./SimpleUpgradable.sol";
```

Taking into account everything stated above the Transpiler given the following input:
```Solidity
pragma solidity ^0.5.0;
import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
import "@openzeppelin/contracts/token/ERC20/ERC20Detailed.sol";

contract GLDToken is ERC20, ERC20Detailed {
    constructor(uint256 initialSupply) public ERC20Detailed("Gold", "GLD", 18) {
        _mint(msg.sender, initialSupply);
    }
}
```

Will produce the following output:

```Typescript
{
 contracts: Array(1) ["GLDToken"],
 fileName: "GLDToken.sol",
 path: "./contracts/__upgradable__/GLDTokenUpgradable.sol",
 source: "..."
}
```
with the following source code value:

```Solidity
pragma solidity ^0.5.0;
import "./@openzeppelin/contracts/token/ERC20/ERC20Upgradable.sol";
import "./@openzeppelin/contracts/token/ERC20/ERC20DetailedUpgradable.sol";
import "@openzeppelin/upgrades/contracts/Initializable.sol";

contract GLDTokenUpgradable is Initializable, ERC20Upgradable, ERC20DetailedUpgradable {
    function initialize(uint256 initialSupply) external initializer {
        __init(true, initialSupply);
    }
    function __init(bool callChain, uint256 initialSupply) internal {
        if(callChain) {
            ContextUpgradable.__init(false);
            ERC20DetailedUpgradable.__init(false, "Gold", "GLD", 18);
        }
        _mint(msg.sender, initialSupply);
    }
}
```Solidity

ylv-io commented:
**Inheritance Call Chain Reconstruction**
Since Solidity compiler can figure out and resolve **Diamon Problem** so we should be able as well. Idea is to reconstruct a full inheritance chain call in the `initializer` method so that would help us to avoid duplicated calls for the ancestors. Example of a transpilied Diamond Problem will look like:

```Solidity
contract C is B1, B2 {
    constructor () B2() B1() public {
        emit Log("C");
    }
}

 contract AU {
    event Log(string);
    
    function initialize() public {
        __init(true);
    }
    
    function __init(bool callChain) internal {
        if(callChain) {
            
        }
        emit Log("A");
    }
}

contract B1U is AU {
    function initialize() public {
        __init(true);
    }
    
    function __init(bool callChain) internal {
        if(callChain) {
            AU.__init(false);
        }
        emit Log("B1");
    }
    
}

contract B2U is AU {
    function initialize() public {
        __init(true);
    }
    
    function __init(bool callChain) internal {
        if(callChain) {
            AU.__init(false);
        }
        emit Log("B2");
    }
}

contract CU is B1U, B2U {
    function initialize() public {
        __init(true);
    }
    
    function __init(bool callChain) internal {
        if(callChain) {
            AU.__init(false);
            B1U.__init(false);
            B2U.__init(false);
        }
        emit Log("C");
    }
}
```

The key implementation detail of this approach is that every contract has two initialize methods: one public `initialize` and with the same parameters as a constructor and second internal `__init` with extra parameter boolean parameter called `callChain` which defines whenever ancestors initializers should be called. The main advantage of this approach compare to idempotency is the gas savings. Extra code does increase the gas cost of the transaction but it can be completely disregarded compare to the costs of the extra `sstore` operations caused by the idempotency approach.

spalladino commented:
Before diving into the code, a comment on *Inheritance Call Chain Reconstruction*. Wouldn't the code be semantically the same if, instead of having a `callChain` flag in the initialization methods, we do something like:

```solidity
contract CU is B1U, B2U {
    function initialize() public {
        AU.__init();
        B1U.__init();
        B2U.__init();
        __init();
    }
    
    function __init() internal {
        emit Log("C");
    }
}
```

ylv-io commented:
@spalladino that seems simple yet brilliant. Well done 🚀It seems to me that it will work yet I would like to finish tests and than try it to run on top of the tests to make sure it works. It is going to be a small change in terms of transpiler source code.

frangio commented on packages/transpiler/src/transformations/fix-import-directives.ts:
```diff
@@ -0,0 +1,20 @@
+    const [start, len] = getSourceIndices(imp);
+    const isTranspiled = artifacts.some(
+      art => art.ast.id === imp.sourceUnit && contracts.some(contract => contract === art.contractName),
+    );
```
I'm not sure that this is correct. Let's say I have a file `A.sol` like this:

```solidity
import './B.sol';
contract A is B { ... }
```

I'm assuming that it would be possible to call `transpile(['A'], [ artifactForA, artifactForB ])`. In that case `B` has to be transpiled because `A` inherits from `B`, but this definition of `isTranspiled` would result in `false` for `import './B.sol'`.

ylv-io replied:
Yep. It is a bit implicit. I've tried to signal that with the name `contractsWithInheritance` but clearly it didn't work. I am open to better name for `contractsWithInheritance`.

frangio commented on packages/transpiler/src/transformations/fix-import-directives.ts:
```diff
@@ -0,0 +1,20 @@
+    return {
+      start,
+      end: start + len,
+      text: !isTranspiled ? `import "${imp.absolutePath}";` : fixedPath,
```
This will not work for complex imports like `import { A } from './A.sol'`. Ideally we would return a transformation that only changes the path string of the import. I would've expected the AST to contain the path string as an AST node but I don't see it.

Additionally, in the case that the import is not transpiled, is it possible to not apply a transformation at all?

ylv-io replied:
Yep. Completely missed `import` syntax. It is also possible to do.
 ```Solidity
import * as symbolName from "filename";
import {symbol1 as alias, symbol2} from "filename";
import "filename" as symbolName;
```

>  I would've expected the AST to contain the path string as an AST node but I don't see it.
It is a property `file` in `ImportDirective`

```JSON
{
        "absolutePath": "contracts/DiamondInheritance.sol",
        "file": "./DiamondInheritance.sol",
        "id": 985,
        "nodeType": "ImportDirective",
        "scope": 1019,
        "sourceUnit": 879,
        "src": "290:60:10"
)
```

ylv-io replied:
There are also other edge cases, for example:
```Solidity
pragma solidity ^0.5.0;

import * as ERC20 from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
import {ERC20Detailed as Detailed} from "@openzeppelin/contracts/token/ERC20/ERC20Detailed.sol";
import "@openzeppelin/contracts/math/Math.sol" as math;
import { DA, DB1, DB2, DC } from "./DiamondInheritance.sol";
import "@openzeppelin/contracts/math/SafeMath.sol";

contract Imports is ERC20.ERC20, Detailed, DC {
    constructor(uint256 initialSupply) public Detailed("Gold", "GLD", 18) {
        uint finalSupply = math.Math.max(0, initialSupply);
        _mint(msg.sender, SafeMath.add(finalSupply, 10));
    }
}
``` 

The whole set of import issues seems big enough to be a separate PR.

frangio commented on packages/transpiler/src/transformations/index.ts:
```diff
@@ -0,0 +1,19 @@
+import { appendDirective } from './append-directive';
```
We can use `export ... from` in this file.

```suggestion
export { appendDirective } from './append-directive';
```

spalladino commented on packages/transpiler/src/transformations/transform-parents.ts:
```diff
@@ -0,0 +1,26 @@
+import { ContractDefinition } from '../solc/ast-node';
+import { Transformation } from '../transformation';
+
+export function transformParents(
```
From what I understand, this transformation only renames the parent names from the inheritance specification (ie from `contract Foo is A, B => contract Foo is AUpgradeable, BUpgradeable`. Shouldn't we also change these names all through the source code, when calling a super function?

spalladino commented on packages/transpiler/src/transformations/build-super-calls-for-chain.ts:
```diff
@@ -0,0 +1,93 @@
+            getContract(contractsToArtifactsMap[base].ast, base),
+            source,
+            contracts,
+            mods,
```
Remember the base contracts can also be initialized when declared in the contract definition, like: `contract Foo is Base(1,2,3) { ... }`. We will need to fetch the parameters from there as well, not just the constructor modifiers.

ylv-io replied:
We do fetch them at the `buildSuperCalls` function using a base class `InheritanceSpecifier`.

frangio commented on packages/transpiler/src/transformations/transform-constructor.ts:
```diff
@@ -0,0 +1,94 @@
+
+  const [start, , contractSource] = getNodeSources(contractNode, source);
+
+  const match = /\bcontract[^\{]*{/.exec(contractSource);
```
>assuming these were not cleared in a previous run by a `transform-parents`

To be clear, my understanding is this would never happen because the transformations are all calculated before being actually applied.

ylv-io replied:
Touché! I haven't accounted for that at all.

frangio commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+        transformations: [
+          appendDirective(art.ast, directive),
+          ...fixImportDirectives(art, artifacts, contractsWithInheritance),
+          ...purgeContracts(art.ast, contractsWithInheritance),
```
>assuming these were not cleared in a previous run by a `transform-parents`

To be clear, my understanding is this would never happen because the transformations are all calculated before being actually applied.

ylv-io commented on packages/transpiler/package.json:
```diff
@@ -0,0 +1,47 @@
+  "main": "./lib/index.js",
+  "scripts": {
+    "test": "npm run test --prefix ./test/solc-0.5",
+    "compile:ts": "rm -rf lib && tsc",
```
No idea, I've just followed `lib` and `cli` on that for consistency. I suggest we address it as a separate PR.

ylv-io commented on packages/transpiler/src/solc/schemas/ast.ts:
```diff
@@ -0,0 +1,38 @@
+  object = { type: 'object' },
+  attrNull = { type: 'null' };
+
+export default {
```
Yes, that is super loose schema. I've added it with a goal in mind to fail early in case if AST files are corrupt for whatever reason. Also, I think we should validate JSON as a general rule rather to trust it to to be what we expect.

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  // create contract name | id to artifact map for quick access to artifacts
+  const contractsToArtifactsMap = artifacts.reduce<Record<string | number, Artifact>>((acc, art) => {
```
>I don't think this is correct if there's more than one contract with the same name in a project.

`oz compile` doesn't support having two contracts with the same name in a project right now. It will produce a corrupt build files if you do that.

How do we expect to handle contracts with the same in a project? What would be their artifacts called? Because right now contract `Counter` has artifact called `Counter.json`.

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+
+  // build a list of all contracts to transpile
+  const contractsWithInheritance = [
+    ...new Set(contracts.map(contract => getInheritanceChain(contract, contractsToArtifactsMap)).flat()),
```
It is sad that vanilla JS support is so bad at node.

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+  });
+
+  // build a array of transformations per Solidity file
+  const fileTrans = contractsWithInheritance.reduce<Record<string, FileTran>>((acc, contractName) => {
```
> The reduce() method executes a reducer function (that you provide) on each element of the array, resulting in a single output value.

`reduce` is far more specific than `for` loop which can be doing anything and it has better semantics for returning single output:
```
const fileTrans = {};
for(...)
```
versus
```
const fileTrans = reduce(...);
```

ylv-io commented on packages/transpiler/src/solc/ast-utils.ts:
```diff
@@ -0,0 +1,118 @@
+  return [start, len, source.slice(start, start + len)];
+}
+
+export function getNode(node: Node, predicate: (node: AnyNode) => boolean): AnyNode | null {
```
Reason is that vars are `undefined` by default in JavaScript so `null` seems more appropriate as something intentionally not having a value compare to `undefined`. What are the reasons to use `undefined` everywhere?

ylv-io commented on packages/transpiler/src/transformations/append-directive.ts:
```diff
@@ -0,0 +1,20 @@
+import { Node } from '../solc/ast-node';
+import { Transformation } from '../transformation';
+
+export function appendDirective(fileNode: Node, directive: string): Transformation {
```
Yes. I want to make transpiled files to look like normal Solidity files so they would be easy to read and audit.

ylv-io commented on packages/transpiler/src/solc/ast-node.ts:
```diff
@@ -0,0 +1,135 @@
+  parameters: VariableDeclaration[];
+}
+
+export type AnyNode = Node | VariableDeclaration | FunctionDefinition | EventDefinition | ModifierDefinition;
```
It allows you to control the list because `AnyNode` is not equal to `T extends Node` and to avoid using generics:
```TypeScript
export function throwIfInvalidNodeTwo<T extends Node>(node: T): void {
``` 
vs
```TypeScript
export function throwIfInvalidNode(node: AnyNode): void {
```
Essentially if you don't want to allow third parties to use derived classes then type list is better than generics.

It also allows you to define user typed type guard but generics doesn't:
```TypeScript
// this works
export function isFunctionDefinition(node: AnyNode): node is FunctionDefinition {
  return node.nodeType === 'FunctionDefinition';
}

// this fails
export function isFunctionDefinition<T extends Node>(node: T): node is FunctionDefinition {
  return node.nodeType === 'FunctionDefinition';
}
```

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  // create contract name | id to artifact map for quick access to artifacts
+  const contractsToArtifactsMap = artifacts.reduce<Record<string | number, Artifact>>((acc, art) => {
```
`fromPairs` is nice but I am fond of vanilla JS.

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+  ].filter(contract => {
+    const art = contractsToArtifactsMap[contract];
+    const contractNode = getContract(art.ast, contract);
+    return isContract(contractNode);
```
`interface` like `IERC20` is not considered a contract.

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+  }, {});
+
+  // build a list of all contracts to transpile
+  const contractsWithInheritance = [
```
Yes, it is misleading. @frangio got tricked by it as well. Great example of bad naming.

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+  }, {});
+
+  // build a final array of files to return
+  return contractsWithInheritance.reduce<OutputFile[]>((acc, contractName) => {
```
Artifacts have duplicated contracts across them so I would have to track which contracts are already transpiled and which not. Seems as almost the same approach.

ylv-io commented on packages/transpiler/src/solc/ast-utils.ts:
```diff
@@ -0,0 +1,118 @@
+  return node.nodeType === name;
+}
+
+export function isImportDirective(node: Node): boolean {
```
That is an amazing note! Wasn't aware of `curry`. My real question why it is called `curry`?

ylv-io commented on packages/transpiler/src/solc/get-inheritance-chain.ts:
```diff
@@ -0,0 +1,11 @@
+
+  return contractNode.linearizedBaseContracts.map(base => {
+    return contractsToArtifactsMap[base].contractName;
+  });
```
Absolutely. You also don't need brackets here `return contractNode.linearizedBaseContracts.map(base => contractsToArtifactsMap[base]);
`. We also can improve eslint config to detect such things.

ylv-io commented on packages/transpiler/src/solc/schemas/ast.ts:
```diff
@@ -0,0 +1,38 @@
+ * Additional attrs can exist depending on the entity the Node represents.
+ */
+
+'use strict';
```
Yes. That was copy pasted from JS project.

ylv-io commented on packages/transpiler/src/solc/schemas/ast.ts:
```diff
@@ -0,0 +1,38 @@
+  type: 'object',
+
+  properties: {
+    nodeType: { type: 'string', minLength: 1 },
```
I am not sure how strict we want to be with JSON validation and how much effort to spent on that. What I am sure is that validating JSON input on some level is a good idea.

ylv-io commented on packages/transpiler/src/transformations/append-directive.ts:
```diff
@@ -0,0 +1,20 @@
+  };
+  const importsAndPragmas = [...getPragmaDirectives(fileNode), ...getImportDirectives(fileNode)];
+  if (importsAndPragmas.length) {
+    const last = importsAndPragmas.slice(-1)[0];
```
Looks nicer.

ylv-io commented on packages/transpiler/src/transformations/fix-import-directives.ts:
```diff
@@ -0,0 +1,20 @@
+    return {
+      start,
+      end: start + len,
+      text: !isTranspiled ? `import "${imp.absolutePath}";` : fixedPath,
```
It doesn't really recreate it here because not every import is using absolute path some are using relative path. Yet there are some cases when transformation is not needed and I addressed them.

ylv-io commented on packages/transpiler/src/transformations/fix-import-directives.ts:
```diff
@@ -0,0 +1,20 @@
+      art => art.ast.id === imp.sourceUnit && contracts.some(contract => contract === art.contractName),
+    );
+    const prefix = !imp.file.startsWith('.') ? './' : '';
+    const fixedPath = `import "${prefix}${imp.file.replace('.sol', 'Upgradable.sol')}";`;
```
Indeed it supports. These are quite the changes so I'll address them as separate PR.

ylv-io commented on packages/transpiler/src/transformations/fix-import-directives.ts:
```diff
@@ -0,0 +1,20 @@
+    const fixedPath = `import "${prefix}${imp.file.replace('.sol', 'Upgradable.sol')}";`;
+    return {
+      start,
+      end: start + len,
```
Yep. I wonder if we will just keep old import side by side with new ones 🤔

ylv-io commented on packages/transpiler/src/transformations/fix-import-directives.ts:
```diff
@@ -0,0 +1,20 @@
+    const isTranspiled = artifacts.some(
+      art => art.ast.id === imp.sourceUnit && contracts.some(contract => contract === art.contractName),
+    );
+    const prefix = !imp.file.startsWith('.') ? './' : '';
```
Since all the transpilied contracts are inside `__upgradeable__` they don't need it in their path because if they import files inside `__upgradeable__` they are fine with relative paths. If it is absolute reference that is also fine. The only case it doesn't cover right now if upgraded contract reference a not upgraded contract from user `contracts` folder. That is related to the `import` issue above. Let me think about that.
![image](https://user-images.githubusercontent.com/3285000/72526062-84134680-38a0-11ea-904e-6224a3964a25.png)

ylv-io commented on packages/transpiler/src/index.ts:
```diff
@@ -0,0 +1,114 @@
+      if (path.startsWith('contracts')) {
+        patchedFilePath = path.replace('contracts/', '');
+      }
+      patchedFilePath = `./contracts/__upgradable__/${patchedFilePath}`;
```
> This places all transpiled contracts in a contracts/__upgradable__ folder, even if the original sources were not in a contracts folder. If the user had all files in the root folder, I'd place the __upgradeable__ folder there as well, instead of creating a new contracts.

Not sure I get it right. You are talking about user having a different folder in a project than `contracts` to store his Solidity contracts?

> And a question: how does the output for contracts imported from node_modules look like here?
`import "./@openzeppelin/contracts/token/ERC20/ERC20Upgradable.sol";` Or you are talking about relative export from `node_modules` more like `import "../node_modules/dep/Foo.sol";`?

ylv-io commented on packages/transpiler/src/transformations/prepend-base-class.ts:
```diff
@@ -0,0 +1,20 @@
+
+  const [start, , nodeSource] = getNodeSources(contractNode, source);
+
+  const regExp = RegExp(`\\bcontract\\s+${contractNode.name}(\\s+is)?`);
```
Do you feel as stripping comments would do the job or there are more edge cases?

ylv-io commented on packages/transpiler/src/transformations/transform-constructor.ts:
```diff
@@ -0,0 +1,94 @@
+    constructorArgsList = constructorNode.parameters.parameters.map(par => par.name).join(',');
+  }
+
+  constructorParameterList = constructorParameterList ? constructorParameterList : '';
```
I love `Nullish Coalescing` and `optional chaining`. Used them quite a bit in C#. I've tried to use them before but our linter complained. Now it works so big thank to the person who fixed it!

ylv-io commented on packages/transpiler/src/transformations/transform-constructor.ts:
```diff
@@ -0,0 +1,94 @@
+  constructorBodySource = constructorBodySource ? constructorBodySource : '';
+  constructorArgsList = constructorArgsList ? `, ${constructorArgsList}` : '';
+
+  const initializeFuncText = isFullyImplemented
```
> I'm assuming that isFullyImplemented means the contract is not abstract (in the solc 0.5 sense).
Correct.
Can't recall at that point but that is wrong because you can call `abstract` method on a child multiple times and that would be an error.

ylv-io replied:
Hmm. I've checked the code. It actually makes sense. Why would you want a public `initialize` method on an abstract class? The real constructor is inside `__init` method.

spalladino replied:
> The real constructor is inside __init method.

Makes sense, since it's `__init` what you call from a derived contract

ylv-io commented on packages/transpiler/src/transformations/build-super-calls-for-chain.ts:
```diff
@@ -0,0 +1,93 @@
+    .flat();
+
+  return [
+    ...new Set(
```
Yes. It is actually [OrderedSet](https://stackoverflow.com/questions/55460303/why-does-js-keep-insertion-order-in-set).

ylv-io commented on packages/transpiler/src/transformations/build-super-calls-for-chain.ts:
```diff
@@ -0,0 +1,93 @@
+            return (constructorNode && !constructorNode.parameters.parameters.length) ||
+              (base.arguments && base.arguments.length)
+              ? buildSuperCall(base.arguments, contractName, source)
+              : [];
```
Not only that but there is also can be a contract with no constructor at all but with filed initializers which will be moved to initializer we have to call. I know think we should always call parent `__init` method.

ylv-io commented on packages/transpiler/src/transformations/purge-var-inits.ts:
```diff
@@ -0,0 +1,19 @@
+    .map(vr => {
+      const [start, , varSource] = getNodeSources(vr, source);
+      const match = /(.*)(=.*)/.exec(varSource);
+      if (!match) throw new Error(`Can't find = in ${varSource}`);
```
The line above `.filter(vr => vr.value && !vr.constant)` picks only variable declarations with assigned values and not constants.

ylv-io commented on packages/transpiler/src/transformations/transform-parents.ts:
```diff
@@ -0,0 +1,26 @@
+          text: `${baseSource}Upgradable`,
+        };
+      });
+  } else return [];
```
Yes. It does.

ylv-io commented on packages/transpiler/src/transformations/transform-parents.ts:
```diff
@@ -0,0 +1,26 @@
+      .filter(base => contracts.some(contract => base.baseName.name === contract))
+      .map(base => {
+        const [start, , baseSource] = getNodeSources(base.baseName, source);
+        const [, len] = getNodeSources(base, source);
```
This are different nodes but they do tend to have the same `src` but I wouldn't rely on that.

ylv-io commented on packages/transpiler/src/transformations/build-super-calls-for-chain.ts:
```diff
@@ -0,0 +1,93 @@
+  const hasInheritance = node.baseContracts.length;
+  if (hasInheritance) {
+    return [
+      ...node.baseContracts
```
Yep. It probably had more expression before.

ylv-io commented on packages/transpiler/src/transformations/build-super-calls-for-chain.ts:
```diff
@@ -0,0 +1,93 @@
+  if (hasInheritance) {
+    return [
+      ...node.baseContracts
+        .filter(base => contracts.some(contract => base.baseName.name === contract))
```
Indeed. It is much cleaner.

ylv-io commented on packages/transpiler/src/transformations/build-super-calls-for-chain.ts:
```diff
@@ -0,0 +1,93 @@
+export function buildSuperCallsForChain(
+  contractNode: ContractDefinition,
+  source: string,
+  contracts: string[],
```
These are all the contracts to transpile. I've renamed them `contractsToTranspile`.

ylv-io commented on packages/transpiler/src/transformations/build-super-calls-for-chain.ts:
```diff
@@ -0,0 +1,93 @@
+  const hasInheritance = node.baseContracts.length;
+  if (hasInheritance) {
+    return [
+      ...node.baseContracts
```
I think no, because here we are collecting parent's constructor calls. Linearized inheritance calls are preserved with `getInheritanceChain` in `buildSuperCallsForChain`.

ylv-io commented on packages/transpiler/src/transformations/transform-parents.ts:
```diff
@@ -0,0 +1,26 @@
+
+        return {
+          start: start,
+          end: start + len,
```
It does.

ylv-io commented:
I've fixed and addressed all the issues raised in reviews by @spalladino and @frangio! Issues which are too big to be addressed as fixes in this PR, I would like to handle as separate PRs. They are:
* [Keep old imports](https://github.com/OpenZeppelin/openzeppelin-sdk/pull/1370#discussion_r365280351)
* Strip comments to prevent [regexp breaks](https://github.com/OpenZeppelin/openzeppelin-sdk/pull/1370#discussion_r365283469)
* Add [old imports](https://github.com/OpenZeppelin/openzeppelin-sdk/pull/1370#discussion_r365289832) to upgraded files
* Check all RegExps against [possible errors](https://github.com/OpenZeppelin/openzeppelin-sdk/pull/1370#discussion_r365306905) with antlr4
* Rename [super calls](https://github.com/OpenZeppelin/openzeppelin-sdk/pull/1370#discussion_r365363762)
* Constructor parent calls may use arguments `constructor(uint z) B(z+1) { }`
